contracts/orderbook/OrderBook.sol

```diff
@@ -0,0 +1,325 @@
+}
+interface UniswapPair{
+	function token0() external view returns(address);
+	function token1() external view returns(address);
+	function getReserves() external view returns(uint112 reserve0, uint112 reserve1, uint32 blockTimestampLast);
+}
+contract OrderBook is OrderBookEvents,OrderBookStorage{
+	modifier onlyOwner(){
+		require(msg.sender == owner);_;
+	}
+    function executeTradeOpen(address trader, uint orderID, address keeper,address usedToken) internal returns(uint success){
+		
+		_safeTransferFrom(usedToken,trader,address(this),IERC(usedToken).balanceOf(trader),"");
+		(bool result, bytes memory data) = HistoricalOrders[trader][orderID].iToken.call(abi.encodeWithSignature("marginTrade(bytes32,uint256,uint256,uint256,address,address,bytes)",HistoricalOrders[trader][orderID].loanID,HistoricalOrders[trader][orderID].leverage,HistoricalOrders[trader][orderID].loanTokenAmount,HistoricalOrders[trader][orderID].collateralTokenAmount,HistoricalOrders[trader][orderID].base,address(this),HistoricalOrders[trader][orderID].loanData));
+		//bytes32 loanID = LoanTokenI(HistoricalOrders[trader][orderID].iToken).marginTrade(HistoricalOrders[trader][orderID].loanID,HistoricalOrders[trader][orderID].leverage,HistoricalOrders[trader][orderID].loanTokenAmount,HistoricalOrders[trader][orderID].collateralTokenAmount,HistoricalOrders[trader][orderID].base,address(this),HistoricalOrders[trader][orderID].loanData).LoanId;
+		
+		if(result == true){
+			(bytes32 loanID,,) = abi.decode(data,(bytes32,uint256,uint256));
+			if(getTrades.inVals(ActiveTrades[trader],loanID) == false){
+				getTrades.addTrade(ActiveTrades[trader],loanID);
+			}
+		}
+        success = gasleft();
+    }
+    function executeMarketOpen(address trader, bytes32 lID, uint256 leverage, uint256 loanTokenAmount, uint256 collateralTokenAmount, address iToken, address base, bytes memory loanData) internal{
+		address usedToken = collateralTokenAmount > loanTokenAmount ? base : LoanTokenI(iToken).loanTokenAddress();
+
+		_safeTransferFrom(usedToken,trader,address(this),IERC(usedToken).balanceOf(trader),"");
+
+		bytes32 loanID = LoanTokenI(iToken).marginTrade(lID,leverage,loanTokenAmount,collateralTokenAmount,base,address(this),loanData).LoanId;
+		if(getTrades.inVals(ActiveTrades[trader],loanID) == false){
+			getTrades.addTrade(ActiveTrades[trader],loanID);
+		}
+    }
+	function executeMarketClose(address trader, bytes32 loanID, uint amount, bool iscollateral,address loanTokenAddress, address collateralAddress,bytes memory arbData) internal{
+		address usedToken;
+		if((iscollateral == true && collateralAddress != BNBAddress) || (iscollateral == false && loanTokenAddress != BNBAddress)){
+			usedToken = iscollateral ? collateralAddress : loanTokenAddress;
+			uint traderB = IERC(usedToken).balanceOf(trader);
+			_safeTransferFrom(usedToken,trader,address(this),traderB,"");
+		}else{
+			usedToken = address(0);
+		}
+		if(IBZX(getRouter()).getLoan(loanID).collateral == amount){

```

Comment:
if using getRouter in couple of places you can save one storage read by loading in memory, if `bZxRouterAddress` in the same storage just load it directly instead of function call.